cranelift/codegen/src/ir/pcc.rs

```diff
@@ -610,118 +1012,186 @@ impl Fact {
+    pub fn union(a: &Fact, b: &Fact) -> Fact {
+        let result = match (a, b) {
             (
-                Fact::DynamicMem {
-                    ty: ty_lhs,
-                    min: min_lhs,
-                    max: max_lhs,
-                    nullable: null_lhs,
+                Fact::Range {
+                    bit_width: bw_lhs,
+                    range: range1,
                 },
-                Fact::DynamicMem {
+                Fact::Range {
+                    bit_width: bw_rhs,
+                    range: range2,
+                },
+            ) if bw_lhs == bw_rhs => Fact::Range {
+                bit_width: *bw_lhs,
+                range: ValueRange::union(range1, range2),
+            },
+
+            (
+                Fact::Mem {
+                    ty: ty_lhs,
+                    range: range1,
+                    nullable: nullable_lhs,
+                },
+                Fact::Mem {
                     ty: ty_rhs,
-                    min: min_rhs,
-                    max: max_rhs,
-                    nullable: null_rhs,
+                    range: range2,
+                    nullable: nullable_rhs,
                 },
-            ) if ty_lhs == ty_rhs && Expr::le(min_rhs, max_lhs) && Expr::le(min_lhs, max_rhs) => {
-                Fact::DynamicMem {
-                    ty: *ty_lhs,
-                    min: Expr::max(min_lhs, min_rhs),
-                    max: Expr::min(max_lhs, max_rhs),
-                    nullable: *null_lhs && *null_rhs,
-                }
-            }
+            ) if ty_lhs == ty_rhs => Fact::Mem {
+                ty: *ty_lhs,
+                range: ValueRange::union(range1, range2),
+                nullable: *nullable_lhs || *nullable_rhs,
+            },
+
+            (
+                Fact::Mem {
+                    ty: ty_mem,
+                    range: range_mem,
+                    nullable: _,
+                },
+                Fact::Range {
+                    bit_width: _,
+                    range: range_offset,
+                },
+            )
+            | (
+                Fact::Range {
+                    bit_width: _,
+                    range: range_offset,
+                },
+                Fact::Mem {
+                    ty: ty_mem,
+                    range: range_mem,
+                    nullable: _,
+                },
+            ) if range_offset.le_expr(&Expr::constant(0)) => Fact::Mem {
+                ty: *ty_mem,
+                range: range_mem.clone(),
+                nullable: true,
+            },
 
             _ => Fact::Conflict,
+        };
+        trace!("Fact::union: {a:?} {b:?} -> {result:?}");
+        result
+    }
+
+    /// Does this fact describe an exact expression?
+    pub fn as_expr(&self) -> Option<&Expr> {
+        match self {
+            Fact::Range {
+                range: ValueRange { equal, .. },
+                ..
+            } => equal.first(),
```

Comment:
Interesting that we can return any of `equal` and be correct here. Does it matter? How are callers using this method? Does it make sense to tweak what is being asked here?